### ISSUE TITLE
Error with ColorPicker 0.16

### ISSUE THREAD

**Issue by MyAnimeManager** [Karakaz/railway-painter#2]
I modified The Color Picker mod to make it work with the 0.16 version (link: old link removed, see answers below). When trying to make a new rule, the game crash with the message
```
 " Error while running event railway-painter::on_gui_click (ID 1)
LuaTable is reporting it's a LuaObject but isn't.
stack traceback:
railway-painter/src/player/gui/color/ColorPicker.lua:17: in function 'setHex'
railway-painter/src/player/gui/color/ColorManager.lua:50: in function 'setHex'
...ay-painter__/src/player/gui/settings/GuiRuleSettings.lua:19: in function 'newRule'
railway-painter/src/player/gui/Gui.lua:27: in function 'newRule'
railway-painter/src/event/EventHandler.lua:52: in function <railway-painter/src/event/EventHandler.lua:40>". 
```
I suspect the problem is the setHex function, in the ColorPicker file, the self.colorPicker isn't a LuaGuiElement but a LuaTable. Do you have any idea on how to solve this ?

**Comment by nobeas-official** [comment]
Thanks I will have a look at it

**Comment by Karakaz** [comment]
Thanks I will have a look at it

**Comment by Karakaz** [comment]
Hmm, seems like the color picker isn't created properly in control.lua add_instance function.

I did a remote call to it right after I received it and got the same error you see
```lua
    self.colorPicker = remote.call("color-picker", "add_instance", {
        parent = parentElement,
        container_name = RPName("color_picker")
    })
    remote.call("color-picker", "set_hex_color", self.colorPicker, "000000")
```

Either the modding interface has changed to break it or your changes did. Did you change anything in that function?

**Comment by MyAnimeManager** [comment]
One of the changes i did was actually an error(removed a "_style"), so i corrected it (new file link: https://mega.nz/#!7xFn3Soa!Q1Rt9rkQMe2d3_0uxmo3t3AoPdcNBjFHUP9fv5Trv24 ), but the error persist. The changes i did were basically to adapt to the GUI changes of 0.16, and were mostly in the style file. I also just checked and the add_instance function is the same, except for the change colspan --> column_count, but i doubt the problem is that. Unless the problem is in the style file, i can't think at anything else, but as i said, i just updated the GUI to follow the new version.

**Comment by MyAnimeManager** [comment]
@Karakaz I checked again all the code, and added a line to see the type of the variable "container" (the one returned by add_instance function). It seems it's indeed returning a table. The strange thing is that clicking the button to open the Color Picker GUI to change the color player doesn't throw the error, even if it's using the exact same function to create his own GUI. Maybe the 0.16 GUI changes broke something? it's strange it doesn't work while you are remote calling it, but it works for itself

**Comment by Karakaz** [comment]
Hey, sorry man its been a busy week for me. I'll look into all this today or over the weekend ;) I'm not ignoring you or anything like that so don't be discouraged :)

**Comment by MyAnimeManager** [comment]
Don't worry, i expected you had something else to do. I was just reporting the thing i found that i thought they could be useful to understand the problem. I'm not in a hurry, so take your time ^^

**Comment by Karakaz** [comment]
Seems like there's a issue with passing the container over the remote interface. I added a printout of the container in color-picker/control.lua:876 and the same in my code, ColorPicker.lua:34:

```lua
log(serpent.block(container))
```
```lua
log(serpent.block(self.colorPicker))
```

and it printed this:
```lua
5656.990 Script @__color-picker__/control.lua:877: {
  __self = "userdata: 0x00007f3eac09cc60"
}
5656.990 Script @__railway-painter__/src/player/gui/color/ColorPicker.lua:34: {
  __self = "userdata: 0x00007f3eac1c60d0"
}
```

Now I'm not sure if it should be different addresses or not but it is suspicious.

If you want to investigate further I suggest you hop on to discord and the modding-help channel and ask about this. And if they don't know then consider creating a bug report to the forums.

I also found this thread which seems to fix a similar bug for 0.16 which may have messed this up https://forums.factorio.com/viewtopic.php?t=54128

**Comment by MyAnimeManager** [comment]
I'm not really an expert on Lua, i just know some things that helped me fixing the GUI error the mod gave, but it's really strange the adresses are different for the same object. I'll try to ask on discord (i suppose you mean the one in the factorio reddit) and let you know if they can understand something more

**Comment by MyAnimeManager** [comment]
Did you use a meta table while calling the remote interface as arguments?

**Comment by Karakaz** [comment]
Nope I haven't touched them. I suspect it's a bug with 0.16 :l An unintended byproduct of the fix in the thread I linked

**Comment by MyAnimeManager** [comment]
Yeah, i spoke with Rsending yesterday, and he confirmed me that, unless a meta table is used, passing a LuaGuiElement should work. He also said they touched the remote interface to fix things, so it's probably a bug from that fix. i'll open a bug report on the forum and let you know

**Comment by Karakaz** [comment]
Great :+1:  Big thank you for putting the work in for us, it can get quite tedious at times...

**Comment by MyAnimeManager** [comment]
Good news. It's was indeed a bug in the game itself, and it will be fixed in the next version

**Comment by MyAnimeManager** [comment]
Just tested, and except for an error in the Color Picker mod that i have to fix (for some reason the "Blue" label isn't show as Blue"), it works pefectly. Since you are reading this, would be possible to support the alpha channel if the ColorPicker is installed?

**Comment by Karakaz** [comment]
Sweeeet! Appreciate the updates :)

My initial thought was that it doesn't make much sense to paint things with transparency which is why it isn't added... I'll git it a try next time I patch it and if it looks decent I will add it mate.

If you get stuck or want some help, give me a shout.

**Comment by MyAnimeManager** [comment]
Well from my experience with the color picker, the transparency help in getting some more "real" color. For example the white, without modifying the alpha is more like a gray, with the alpha you can get a real white.

**Comment by MyAnimeManager** [comment]
Just a little heads up. I had to change the remote interface name to "color-picker16" for the updated version of the mod. So to make it works, you should probably change it in the remote call and in the ColorManager, so it can actually see the ColorPicker ^^
